### PR TITLE
Basic D3 Result Graph

### PR DESCRIPTION
**Pull request by bfbachmann**
## Related Issue
#5  

## Description
* Added a Resultgraph that creates Nodes and Edges from the search results it's given and plots them. It will update every time the user performs a search.
* Renamed `SearchResult` to `ResultItem` to make it more clear that the component is only for use in the `ResultList`.

## WIKI Updates
* None 😬 

## Todos
* Update this PR when the API changes on the front and back ends have been merged.
* Fix up code smells all over the place 😢 

General:
- [ ] Tests
- [X] Documentation
- [ ] Wiki

Other (links to TODOs in code):
* [don't make all edge weights 3](https://github.com/ubclaunchpad/sleuth-frontend/compare/d3-graph?expand=1#diff-823a46cdb4ff584bec6bb4aa1d2eaf3dR135)
* [add categorization](https://github.com/ubclaunchpad/sleuth-frontend/compare/d3-graph?expand=1#diff-823a46cdb4ff584bec6bb4aa1d2eaf3dR232)

### CODE REVIEW

**Comment by bobheadxi** [comment]
@bfbachmann After we merge #6 it would be cool if we could update this PR and add node edges between [`courseItem`](https://github.com/ubclaunchpad/sleuth-frontend/pull/6/files#diff-285c8b6eee5742bbe36d9b94a856e0f0R79)'s by [`subjectID`](https://github.com/ubclaunchpad/sleuth/blob/master/sleuth_backend/solr/models.py#L71) (ie link all `courseItem` that share the same `subjectID`), and maybe even generate a new node for the parent subject using `subjectID` and [`subjectData`](https://github.com/ubclaunchpad/sleuth/blob/master/sleuth_backend/solr/models.py#L72):
```js
...
'subjectID': <SUBJECT_URL>,
'subjectData': [
    <SUBJECT_CODE>,
    <SUBJECT_TITLE>,
],
...
```
Speaking of which, I should probably note what `subjectData` looks like in the documentation oops

File: src/components/ResultGraph.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                .on('start', this._startDrag)
+                .on('drag', this._doDrag)
+                .on('end', this._endDrag));
+
+        // Each node gets a title based on it's id (url).
+        this.nodes.append('title')
+            .text(node => { return node.id; });
+
+        // Add forces on the edges between nodes to pull them back together when
+        // they are separated.
```
Comment (bobheadxi) [inline src/components/ResultGraph.js:75]: documentation is A+

File: src/components/SearchForm.js [hunk 2]
```diff
@@ -59,36 +68,47 @@ export default class SearchForm extends React.Component {
...
-                case "courseItem":
+                case 'courseItem':
                     results = results.concat(docs.map(doc => {
-                        return {
-                            url: doc.id,
-                            description: doc.description,
-                            pageName: doc.name,
-                            siteName: doc.subjectData.length == 2 ? doc.subjectData[1] : ''
-                        }
+                        return new CourseItemResult(
```
Comment (bobheadxi) [inline src/components/SearchForm.js:96]: 🔥

File: src/components/ResultGraph.js [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        }
+
+        this._createColor = scaleOrdinal(schemeCategory20);
+        this._tick = this._tick.bind(this);
+        this._startDrag = this._startDrag.bind(this);
+        this._endDrag = this._endDrag.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.queryId === this.props.queryId) {
```
Comment (bobheadxi) [inline src/components/ResultGraph.js:30]: niceee much better than `stringify`

**Review by bobheadxi — APPROVED** [review]
amaze

File: src/components/ResultTypes.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+     * Creates a CourseItemResult from the given search result data
+     * @param {String} id
+     * @param {String} description
+     * @param {String} pageName
+     * @param {String} siteName
+     */
+    constructor(id, description, pageName, siteName) {
+        this.url = id;
+        this.description = description;
+        this.pageName = name;
```
Comment (bobheadxi) [inline src/components/ResultTypes.js]: This should be `this.pageName = pageName` right

**Review by bobheadxi — CHANGES_REQUESTED** [review]

File: src/components/ResultTypes.js (hunk 4 above)
Comment (bfbachmann) [reply src/components/ResultTypes.js]: Oh damn... I wonder why that didn't produce an error. I guess it just does `this.pageName = undefined`.

**Review by bobheadxi — APPROVED** [review]

**Pull request merged**